### ISSUE TITLE
Occasional NPE when calling JupyterConnection.connect

### ISSUE THREAD

**Issue by dgelessus** [SpencerPark/jupyter-jvm-basekernel#9]
Sometimes when my kernel's `main` method calls `JupyterConnection.connect`, I get a NPE and the kernel fails to connect to Jupyter. Here's the log output (including stack trace) when this happens:

```
Jun 06, 2018 2:39:45 PM io.github.spencerpark.jupyter.channels.HeartbeatChannel bind
INFORMATION: Polling on Heartbeat-0
Jun 06, 2018 2:39:45 PM io.github.spencerpark.jupyter.channels.ShellChannel lambda$bind$17
INFORMATION: Shell-0 shutdown.
Exception in thread "main" java.lang.NullPointerException
        at io.github.spencerpark.jupyter.channels.ShellChannel.bind(ShellChannel.java:68)
        at io.github.spencerpark.jupyter.channels.JupyterConnection.lambda$connect$15(JupyterConnection.java:48)
        at io.github.spencerpark.jupyter.channels.JupyterConnection.forEachSocket(JupyterConnection.java:74)
        at io.github.spencerpark.jupyter.channels.JupyterConnection.connect(JupyterConnection.java:48)
        at de.prob2.jupyter.Main.main(Main.java:41)
[W 14:39:47.091 NotebookApp] Timeout waiting for kernel_info reply from 256259e8-8ac5-49ca-8935-b11acab9cc23
```

I'm using version 2.2.0-SNAPSHOT, but this has happened on previous versions as well. It's not common enough to be a major problem, but it does happen regularly every now and then. When it happens, restarting the kernel is usually enough to fix it.

**Comment by SpencerPark** [comment]
You shouldn't get the NPE any more but there still may be an underlying problem as the NPE should only occur if the channel is closed almost immediately after opening. If you find this is still a problem please reopen this issue as it is related. 

I have a feeling this is related to a development workflow replacing the jar and constantly restarting but I'm not sure how to reproduce it. I'll keep an eye out myself as well.

**Comment by dgelessus** [comment]
> I have a feeling this is related to a development workflow replacing the jar and constantly restarting

That is more or less what I'm doing. (I have my kernelspec pointing to the jar in the Gradle build output directory.) Previously I left the kernel running while rebuilding the jar, and that caused some really odd issues, like the JVM crashing during the build. Now I always shut the kernel down before I rebuild the jar, which has fixed the JVM crashes, but not the NPEs.

Thanks for the fix, I'll see how things work with the new version.

**Comment by dgelessus** [comment]
You were correct, the underlying issue is still there. At least now it doesn't cause any exceptions, but the kernel still fails to start up. Here's how the log typically looks when it happens:

```
[I 18:10:04.627 NotebookApp] Kernel started: 557f41f1-ef28-4922-ac21-093f7a9c476a
Listening for transport dt_socket at address: 54966
[W 18:10:14.778 NotebookApp] Timeout waiting for kernel_info reply from 557f41f1-ef28-4922-ac21-093f7a9c476a
Jun 20, 2018 6:10:14 PM io.github.spencerpark.jupyter.channels.HeartbeatChannel bind
INFORMATION: Polling on Heartbeat-0
Jun 20, 2018 6:10:14 PM io.github.spencerpark.jupyter.channels.HeartbeatChannel lambda$bind$4
INFORMATION: Heartbeat-0 shutdown.
Jun 20, 2018 6:10:14 PM io.github.spencerpark.jupyter.channels.ShellChannel lambda$bind$17
INFORMATION: Shell-0 shutdown.
Jun 20, 2018 6:10:14 PM io.github.spencerpark.jupyter.channels.ShellChannel bind
INFORMATION: Polling on Shell-0
Jun 20, 2018 6:10:14 PM io.github.spencerpark.jupyter.channels.ShellChannel lambda$bind$17
INFORMATION: Shell-1 shutdown.
Jun 20, 2018 6:10:14 PM io.github.spencerpark.jupyter.channels.ShellChannel bind
INFORMATION: Polling on Shell-1
```

I don't know if there's anything specific that triggers it, so far it seems to be basically random. I'm fairly sure that it's *not* related to rebuilding the kernel jar while it's running, I've also had this happen with notebooks that I just opened.

I know this isn't a lot of information - if there's anything else I can provide to help you debug this, let me know. (I have my kernel set up so I can connect with a debugger, if that's of any use...)

**Comment by SpencerPark** [comment]
Thanks for the update. 

I tried to reproduce as well as look into jeromq to see if maybe there is an issue on that end but to no avail. I've looked at issues with the debugger port colliding, tried running multiple kernels, stress testing reloads, explicit port collisions, nothing is reproducing the error.

The client isn't explicitly closing the connection as we would have a log entry about that shutdown_request. The ordering of the shutdowns has me convinced that JupyterConnection#close is what is triggering it but it shouldn't be.

If I may ask, what OS, jvm, python, and jupyter versions are you on? I've published a version with some extra logging lines which might be helpful and running jupyter with the `--debug` flag could also get some more information.

**Comment by dgelessus** [comment]
It just happened again, and this was with a freshly started Jupyter and the first notebook I opened. That confirms that it's not because of me rebuilding the kernel.

I'm using Mac OS X 10.11.6, Oracle JDK 8u172, Python 3.6.4, jupyter-core 4.4.0, jupyter-client 5.2.3. Here's the console output with `--debug` (I removed all the lines about successful HTTP requests and such):

```
[I 10:45:23.879 NotebookApp] Kernel started: a9dab926-0961-4844-9446-e3c223a3a0de
[D 10:45:23.880 NotebookApp] Kernel args: {'kernel_name': 'prob2', 'cwd': '/Users/Shared/Uni/SHK/ProB2/prob2-jupyter-kernel/notebooks'}
[D 10:45:23.886 NotebookApp] 201 POST /api/sessions (::1) 116.17ms
[D 10:45:24.061 NotebookApp] Using contents: services/contents
Listening for transport dt_socket at address: 49551
[D 10:45:24.566 NotebookApp] Initializing websocket connection /api/kernels/a9dab926-0961-4844-9446-e3c223a3a0de/channels
[D 10:45:24.571 NotebookApp] Requesting kernel info from a9dab926-0961-4844-9446-e3c223a3a0de
[D 10:45:24.571 NotebookApp] Connecting to: tcp://127.0.0.1:49540
[W 10:45:34.576 NotebookApp] Timeout waiting for kernel_info reply from a9dab926-0961-4844-9446-e3c223a3a0de
[D 10:45:34.578 NotebookApp] Opening websocket /api/kernels/a9dab926-0961-4844-9446-e3c223a3a0de/channels
[D 10:45:34.579 NotebookApp] Getting buffer for a9dab926-0961-4844-9446-e3c223a3a0de
[D 10:45:34.579 NotebookApp] Connecting to: tcp://127.0.0.1:49540
[D 10:45:34.580 NotebookApp] Connecting to: tcp://127.0.0.1:49541
[D 10:45:34.580 NotebookApp] Connecting to: tcp://127.0.0.1:49542
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.HeartbeatChannel bind
INFORMATION: Binding Heartbeat-0 to tcp://127.0.0.1:49544.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop start
INFORMATION: Loop starting...
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop start
INFORMATION: Loop started.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.HeartbeatChannel bind
INFORMATION: Polling on Heartbeat-0
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.ShellChannel bind
INFORMATION: Binding Shell-0 to tcp://127.0.0.1:49540.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop start
INFORMATION: Loop starting...
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop run
INFORMATION: Running loop shutdown callback.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.ShellChannel lambda$bind$17
INFORMATION: Shell-0 shutdown.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop start
INFORMATION: Loop started.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop run
INFORMATION: Loop stopped.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.ShellChannel bind
INFORMATION: Polling on Shell-0
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.ShellChannel bind
INFORMATION: Binding Shell-1 to tcp://127.0.0.1:49543.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop start
INFORMATION: Loop starting...
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.Loop start
INFORMATION: Loop started.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.ShellChannel bind
INFORMATION: Polling on Shell-1
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.StdinChannel bind
INFORMATION: Binding stdin to tcp://127.0.0.1:49542.
Jun 21, 2018 10:45:35 AM io.github.spencerpark.jupyter.channels.IOPubChannel bind
INFORMATION: Binding iopub to tcp://127.0.0.1:49541.
```

I actually have a theory what might be happening here. Based on the log, we know that:

* Because of the "Running loop shutdown callback" message, the `while (running)` loop in `Loop.run` must have exited, i. e. `running` must be `false`.
* `running` is set to `false` in three places: in the initializer, the `InterruptedException` handler, and `shutdown`.

https://github.com/SpencerPark/jupyter-jvm-basekernel/blob/53bc4372de0ab2fa6c42299dc8273deb2f05da24/src/main/java/io/github/spencerpark/jupyter/channels/Loop.java#L10

https://github.com/SpencerPark/jupyter-jvm-basekernel/blob/53bc4372de0ab2fa6c42299dc8273deb2f05da24/src/main/java/io/github/spencerpark/jupyter/channels/Loop.java#L53-L56

https://github.com/SpencerPark/jupyter-jvm-basekernel/blob/53bc4372de0ab2fa6c42299dc8273deb2f05da24/src/main/java/io/github/spencerpark/jupyter/channels/Loop.java#L77-L80

* The last two cases produce log messages, but neither of them appear in the log. That means that `running` was initialized to `false` when the `Loop` was created, and wasn't set to `true` before the `while (running)` loop started.
* `running` is unconditionally set to `true` in `Loop.start`.

https://github.com/SpencerPark/jupyter-jvm-basekernel/blob/53bc4372de0ab2fa6c42299dc8273deb2f05da24/src/main/java/io/github/spencerpark/jupyter/channels/Loop.java#L70-L75

* This means that `Loop.run` must have run before the `this.running = true` line. This is entirely possible, because `super.start()` starts a new thread, which might get to run immediately, before `this.running = true` is executed.

So we have a race condition: depending on how the scheduler is feeling, `running` may or may not be set to `true` before the `while (running)` loop is reached.

I think the easiest fix would be to move the `this.running = true` line from `start` into `run`, so that the assignment happens in the newly started thread rather than the parent thread.

**Comment by SpencerPark** [comment]
> So we have a race condition: depending on how the scheduler is feeling, running may or may not be set to true before the while (running) loop is reached.

Uggh yes a race condition makes perfect sense. I'm a little embarrassed of this one :). Thank you for the detailed outline of the problem and for your help debugging it. 

> I think the easiest fix would be to move the this.running = true line from start into run, so that the assignment happens in the newly started thread rather than the parent thread.

Thinking about what you mentioned about the scheduler, if we set `running` to `true` in `run()` and for some reason (a real reason this time) the loop is `shutdown()` before the tread gets up and running then the `shutdown()` request gets ignored. So I just moved the assignment to right before the `super.start()` which is where it should have been in the first place. I'm hesitant to make things any more complicated in case I introduce another bug like this but we really need `running` to have 3 states. Anyways I don't want to touch it as things should be working now. Thank you again and the fix will be published immediately.
